### PR TITLE
feat: custom refresh frequency

### PR DESCRIPTION
**Pull request by Abhishek-kumar-ISM**
Earlier at dashboard level we had very limited options in 'Set auto-refresh interval' as shown in image below 

107 99 45 216_9000_superset_dashboard_5__native_filters_key=6ZOG4aE3VjGvXnNg9wOkgTrlcVVgk1AgOh1wN3MlrGFIXarapEImbmrarIlIKHQq (3)

normal

But as per our organisation level we wanted the frequency to be something else that was not in dropdown, so we implemented that.

As part of this PR, we can add custom frequency as required by user, user can select any frequency in hour, minute and seconds and set it as auto refresh interval frequency (in seconds).

107 99 45 216_9000_superset_dashboard_5__native_filters_key=6ZOG4aE3VjGvXnNg9wOkgTrlcVVgk1AgOh1wN3MlrGFIXarapEImbmrarIlIKHQq

107 99 45 216_9000_superset_dashboard_5__native_filters_key=6ZOG4aE3VjGvXnNg9wOkgTrlcVVgk1AgOh1wN3MlrGFIXarapEImbmrarIlIKHQq (2)

107 99 45 216_9000_superset_dashboard_5__native_filters_key=6ZOG4aE3VjGvXnNg9wOkgTrlcVVgk1AgOh1wN3MlrGFIXarapEImbmrarIlIKHQq (1)

### CODE REVIEW

**Comment by Abhishek-kumar-samsung** [comment]
@rusackas @justinpark @mistercrunch @villebro @geido @eschutho @betodealmeida @nytai @craig-rueda @kgabryje @dpgaspar 

Please review.

**Comment by john-bodley** [comment]
@kasiazjc would you mind looking at the design? Personally I think that custom (and how it is defined/configured) should be a secondary option (hidden by default)—possibly under a collapsed menu, button, tab, etc.

**Comment by craig-rueda** [comment]
From the screenshots, it looks a little confusing to show the dropdown alongside the text fields. Perhaps a new "virtual" option (`CUSTOM`) should be added, which would then hide/show the text boxes when selected? Also, having two buttons when choosing a custom refresh interval looks a little confusing as well.

**Comment by Abhishek-kumar-samsung** [comment]
Okay sure @john-bodley @craig-rueda 
I will try to hide custom view and show custom view when required.

**Comment by Abhishek-kumar-samsung** [comment]
I have made changes as asked
When user will click **CUSTOM** button then box changes so that one can provide custom frequency, and if user clicks **CUSTOM** button again then original window will appear again.

107 99 45 216_9000_superset_dashboard_5__native_filters_key=91ksfxXf42UqpEDLp_k4NNvRNLAqehP-I5U_tTUR-qOzjqVLz4Xr-SWMRYh8cGHh (1)

107 99 45 216_9000_superset_dashboard_5__native_filters_key=91ksfxXf42UqpEDLp_k4NNvRNLAqehP-I5U_tTUR-qOzjqVLz4Xr-SWMRYh8cGHh

**Comment by Abhishek-kumar-samsung** [comment]
@john-bodley @craig-rueda Please check and review.

I have shown only one refresh change option at a time as requested by @john-bodley and removed extra submit button as requested by @craig-rueda

**Comment by rusackas** [comment]
Sorry for the churn and delays on this, but I think the main issue is UI design, not the technical implementation.

My own opinion:
• It would be nice if it had the Select menu (including the Custom option, rather than a button for Custom)
• If (and only if) the Custom option is selected from the Select menu, would the additional fields appear (or disappear)
• The Select menu would remain present... if you select something other than Custom, the form elements are removed, and the values ignored. 

As a side note, I feel like the Submit button should be on the right, and the cancel button on the left. I'm not sure if we have a solid documented pattern around this, though (cc. @kasiazjc)

**Comment by Shisir99** [comment]
@Abhishek-kumar-samsung is this frequency only for a session or persistent?

**Comment by Abhishek-kumar-samsung** [comment]
> Sorry for the churn and delays on this, but I think the main issue is UI design, not the technical implementation.
> 
> My own opinion: • It would be nice if it had the Select menu (including the Custom option, rather than a button for Custom) • If (and only if) the Custom option is selected from the Select menu, would the additional fields appear (or disappear) • The Select menu would remain present... if you select something other than Custom, the form elements are removed, and the values ignored.
> 
> As a side note, I feel like the Submit button should be on the right, and the cancel button on the left. I'm not sure if we have a solid documented pattern around this, though (cc. @kasiazjc)

Okay i will do some modification with UI.

**Comment by Abhishek-kumar-samsung** [comment]
> @Abhishek-kumar-samsung is this frequency only for a session or persistent?

It is persistent.

**Comment by Shisir99** [comment]
@Abhishek-kumar-samsung my dashbaord frequency is saving only for a particular session , how can i make it persistent?

**Comment by kasiazjc** [comment]
> > Sorry for the churn and delays on this, but I think the main issue is UI design, not the technical implementation.
> > My own opinion: • It would be nice if it had the Select menu (including the Custom option, rather than a button for Custom) • If (and only if) the Custom option is selected from the Select menu, would the additional fields appear (or disappear) • The Select menu would remain present... if you select something other than Custom, the form elements are removed, and the values ignored.
> > As a side note, I feel like the Submit button should be on the right, and the cancel button on the left. I'm not sure if we have a solid documented pattern around this, though (cc. @kasiazjc)
> 
> Okay i will do some modification with UI.

Thanks and sorry for the long wait! What I'm thinking in terms of UI improvements is this: 

-> instead of button/switcher use single select, for selecting the type of interval, the options are:

- don’t refresh

- custom interval (triggers the additional fields as shown below)

- all of the preset options that are available today (in the example dropdown I listed only two, but should be all of them)

   
-> set fixed height for the modal to accommodate for disappearing custom fields, so that the height does not jump

image

I think this way setting up the custom interval will be more intuitive, as we are using common patterns. What do you think and could you possibly take this up @Abhishek-kumar-samsung ?

**Comment by Abhishek-kumar-samsung** [comment]
@kasiazjc 

Yes i will take this, i will try to make as you have suggested.
I was busy in last few days, so i will work on this from now.

**Comment by Abhishek-kumar-samsung** [comment]
> @Abhishek-kumar-samsung my dashboard frequency is saving only for a particular session , how can i make it persistent?

@Shisir99 i am sorey, i was not sure that time, yes it is not persistent for me too, that is for a particular session only.

@rusackas isn't that a problem? 
Automatic refresh interval for just a single session does not makes much sense(according to me, or maybe i missed something), it should be persistent.

**Comment by Abhishek-kumar-samsung** [comment]
dont_refresh
custom
10_sec


@kasiazjc i implemented the design as you have suggested, Please review.

**Comment by Abhishek-kumar-samsung** [comment]
@rusackas @justinpark @mistercrunch @villebro @geido @eschutho @betodealmeida @nytai @craig-rueda @kgabryje @dpgaspar @kasiazjc @john-bodley 

Please review.

**Comment by kasiazjc** [comment]
> dont_refresh custom 10_sec
> 
> @kasiazjc i implemented the design as you have suggested, Please review.

Thank you so much! The only thing I would suggest is changing minutes + seconds to a searchable select dropdown component, so that the numbers for seconds (59) and minutes (59) are fixed. Could be a weird experience if people would type in for example 1 hour, 72 minutes, 342 seconds, so I think to avoid that selects would help.

**Comment by Abhishek-kumar-samsung** [comment]
Hi @kasiazjc 
I added dropdowns in minute and seconds as you have suggested.
Instead of select i used input type with datalist.
In datalist i have provided options from 0-59, if user puts some other value and submits then it will display error message.

datalist is looking little different depending on browsers, some previews as below.

**edge browser preview:**
IMG_0831

**firefox browser preview:**
IMG_0830

**chrome browser preview:**
IMG_0829

I was having problem taking screenshot with dropdown, so i took screenshot using phone.

Please check and review @kasiazjc @rusackas

**Comment by Abhishek-kumar-samsung** [comment]
@kasiazjc @rusackas can you pls check the PR, there's been a long time since no comments, so i thought to remind.

**Comment by LyleScott** [comment]
This would be a welcome addition, thanks for the effort.

**Comment by Abhishek-kumar-samsung** [comment]
@kasiazjc @rusackas can you pls check the PR, there's been a long time since nobody reviewed it.
If any problem is there then pls tell.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (eschutho) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:179]: Hi, instead of using css to hide/show can we use React to conditionally render the element?

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (eschutho) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:95]: is it possible to control/select this element with a ref instead of directly fetching from the dom?

**Comment by kasiazjc** [comment]
> Hi @kasiazjc I added dropdowns in minute and seconds as you have suggested. Instead of select i used input type with datalist. In datalist i have provided options from 0-59, if user puts some other value and submits then it will display error message.
> 
> datalist is looking little different depending on browsers, some previews are as below.
> 
> **edge browser preview:** IMG_0831
> 
> **firefox browser preview:** IMG_0830
> 
> **chrome browser preview:** IMG_0829
> 
> I was having problem taking screenshot with dropdown, so i took screenshot using phone.
> 
> Please check and review @kasiazjc @rusackas

Sorry for the super late review! Missed this PR. These are not ant design components right? I think single select with predefined list in the dropdown we have in code should be the one used in this for consistency (design and patterns). You can find it in filters for example, and users can also type in some numbers (it just would not accept the new number)

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:188]: I am changing views based on dropdown selected options, i was trying by using react but it was not working,
I am not sure about this, but i will check and try to do if possible.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:101]: while implementing this, i was not getting how to hide/unhide custom options from react so i implemented onclick to hide/unhide dom element.

Not sure how can i do without dom element, but i will check and do if possible.

**Comment by Abhishek-kumar-samsung** [comment]
> Sorry for the super late review! Missed this PR. These are not ant design components right? I think single select with predefined list in the dropdown we have in code should be the one used in this for consistency (design and patterns). You can find it in filters for example, and users can also type in some numbers (it just would not accept the new number)

@kasiazjc I have made changes as per your suggestion, i have removed and put select dropdown, now one can also filter based on select dropdown options, now it looks exactly similar to earlier dropdowns, i have checked in chrome, edge, firefox, it is similar in all browsers.

**Comment by Abhishek-kumar-samsung** [comment]
IMG_1487

**Comment by kasiazjc** [comment]
> IMG_1487

This looks great, thank you! One last request - could you also change "hours" just to normal input? I think there doesn't need to be any limit and user can type in any number. And the same with styling - if you could use input field that we already have in the code that would be great :)

**Comment by Abhishek-kumar-samsung** [comment]
@kasiazjc hours was normal input text only, it can take any numeric value greater than or equal to 0. 
Regarding styling of input text, i removed my custom styling and took one classname from other file and put it.(i.e used styling that was already in the code)

Please check.

**Comment by Abhishek-kumar-samsung** [comment]
@kasiazjc @rusackas pls review, its long time since no reply.

**Comment by rusackas** [comment]
I think it's close, but the main concerns remain that it's fetching the element from the dom to manipulate it (rather than a ref) and using CSS to show/hide, rather than conditional rendering. It seems like it's working _around_ react, rather than leveraging React. If that can be fixed up, I think it'll be good to go.

**Comment by Abhishek-kumar-samsung** [comment]
> I think it's close, but the main concerns remain that it's fetching the element from the dom to manipulate it (rather than a ref) and using CSS to show/hide, rather than conditional rendering. It seems like it's working _around_ react, rather than leveraging React. If that can be fixed up, I think it'll be good to go.

Okay sure, thankyou for review.
I will correct it and then ask again for review.

**Comment by geido** [comment]
Hello @Abhishek-kumar-samsung just checking in here. Let me know if you need any support to get this to the finish line. Thanks for your contribution!

**Comment by Abhishek-kumar-samsung** [comment]
Hello @rusackas , @geido , @eschutho  , i have removed the use of DOM element to hide/show custom view. 
Now i have made an element custom_block whose state is determining the visibility of custom view.

Please review.

**Comment by geido** [comment]
@Abhishek-kumar-samsung would you please pull master in your branch and push back?

**Comment by Abhishek-kumar-samsung** [comment]
> @Abhishek-kumar-samsung would you please pull master in your branch and push back?

Done, now it is updated to latest master branch

**Comment by geido** [comment]
Hello @Abhishek-kumar-samsung I am observing some behaviours that should be fixed. In the video you can see the following:

- When going from an option to another, the initial values that I did set for the custom options are not cleared
- When moving from a custom option to a standard option, the values will stay on screen for a short blink and then disappear


https://github.com/apache/superset/assets/60598000/cb53640a-b8e0-4580-a22b-a41f662be3bb

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:112]:
```
this.setState({
  custom_block: value === -1,
});
```

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:122]:
```suggestion
    if (!refreshIntervalOptions.length) {
```

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:137]: I think you need to review these logics here. Are you trying to put the custom interval at a specific position of the array?

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:147]:
```
min_sec_options(min_or_sec: string) {
  return Array.from({ length: 60 }, (_, i) => ({
    value: i,
    label: `${i} ${min_or_sec}`,
  }));
}
```

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:192]: What is that is not working specifically? We definitely need to show this conditionally and not hidden via CSS

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:199]: Please use the Input provided by Antdesign in src/components

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:204]: This needs to be localized `t('Type a number')`

**Review by geido — CHANGES_REQUESTED** [review]
Thanks for the contribution! There are some important changes required from a first-pass review. Let me know when you are ready for another round and I will be happy to have another look

**Comment by Abhishek-kumar-samsung** [comment]
Thankyou for reviews @geido 
I will check and correct the things.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:137]: Yes i kept custom options in the array options that we are showing in dropdown.

Comment (geido) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:137]: Ok then please consider using `map` rather than the for loop to make the code more readable and concise. Thank you

**Comment by maanwater** [comment]
Do you think this will make it to the 4.0.0 release?

**Comment by Abhishek-kumar-samsung** [comment]
> Thanks for the contribution! There are some important changes required from a first-pass review. Let me know when you are ready for another round and I will be happy to have another look

@geido 
Can you pls have another look, i resolved all the things that you had mentioned.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:130]: The label needs to be localized. Also, aren't you adding this two times as you are adding it when the above condition triggers too?

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:185]: Not sure why we are still resorting to style visibility rather than rendering conditionally?

**Review by geido — COMMENTED** [review]
Sorry to be pedantic here but I still see a few issues. Also, just realized this is missing tests which are required for the PR to get merged in. Let me know if you need any additional clarification or help, you can find me on Apache Superset Slack

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:130]: Ok i will localize it,
no i am not adding twice because from if condition i am returning.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:185]: that custom block visibility depends on whether the custom interval option is selected or not, what else can be done to render it conditionally?
from react i am changing the state of custom_block flag, and if it is true then the block gets rendered else not.

Comment (geido) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:185]: What I am trying to say is that you should remove the element rather than hiding it with opacity.

See for example `{showRefreshWarning && (...)}` a few lines below.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (geido) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:130]: What I mean is that you are duplicating the code by using the condition `!refreshIntervalOptions.length` and then pushing the custom interval option, but you also do the same when `refreshIntervalOptions.length` is larger than 0.

So just do:

```
refresh_options.push({ value: -1, label: 'Custom interval' });

if (refreshIntervalOptions.length) {
   refresh_options.push(
      ...refreshIntervalOptions.map(option => ({
        value: option[0],
        label: t(option[1]),
      })),
    );
}

return refresh_options;
```

**Comment by Abhishek-kumar-samsung** [comment]
@geido can you pls review again :)
I have made modifications as asked by you.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:265]: It would be better if handlers called an external function rather than defining the logic inline.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx:185]: We prefer not to use inline styles, or hard-coded pixel values. We prefer using Emotion (kind of like Styled Components) and using the Theme's gridUnits. There are some style guides available [here](https://github.com/apache/superset/wiki/Emotion-Styling-Guidelines-and-Best-Practices).

**Comment by coetzeevs** [comment]
Hi all - please direct me to an issue link if there's a separate place for questions. We recently adopted Superset and I'm looking into setting up refresh schedules for our dashboards. This PR comes close to introducing the feature I think we need, but really something with the flexibility like cron schedule expressions would be ideal. I haven't been able to understand when the refresh interval triggers (i.e. how it's determined when it'll happen) or to specify a time it should kick off. 

Effectively I just want to specify when my refresh interval should start and (based off the time window I choose) when it should stop. Can any help me understand this, please?

**Comment by Abhishek-kumar-samsung** [comment]
> Hi all - please direct me to an issue link if there's a separate place for questions. We recently adopted Superset and I'm looking into setting up refresh schedules for our dashboards. This PR comes close to introducing the feature I think we need, but really something with the flexibility like cron schedule expressions would be ideal. I haven't been able to understand when the refresh interval triggers (i.e. how it's determined when it'll happen) or to specify a time it should kick off.
> 
> Effectively I just want to specify when my refresh interval should start and (based off the time window I choose) when it should stop. Can any help me understand this, please?

Currently in this PR, i was implementing refresh intervals based on relative timing. But crons idea would be nice.

If we need to include crons then this PR will extend to some long time, maybe endlessly.

So maybe we can start an SIP with multiple stages of PR, 
First stage maybe this one i.e relative time, 
and Second stage can be crons.

What say? @rusackas

**Comment by rusackas** [comment]
I agree we don't want to grow the scope of this PR. We should merge this first, and then deal with the cron idea (which is a nice idea) as a subsequent PR. 

For the cron feature, it seems like a reasonable plan to seek "lazy consensus" on the dev@ list. Then if that starts an argument, we can upgrade the conversation to a SIP and vote it.

**Comment by Abhishek-kumar-samsung** [comment]
> I agree we don't want to grow the scope of this PR. We should merge this first, and then deal with the cron idea (which is a nice idea) as a subsequent PR.
> 
> For the cron feature, it seems like a reasonable plan to seek "lazy consensus" on the dev@ list. Then if that starts an argument, we can upgrade the conversation to a SIP and vote it.

Yes sure, i will quickly do the two changes that you had suggested and once merged, we will raise SIP for cron.

Then we can work on it, @coetzeevs you can also work on that if you had implemented something in it.

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx]: Done

File: superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx
Comment (Abhishek-kumar-samsung) [reply superset-frontend/src/dashboard/components/RefreshIntervalModal.tsx]: Done

**Comment by Abhishek-kumar-samsung** [comment]
@geido @rusackas i resolved whatever issues pointed out, can you pls review.

**Comment by LeoDiep** [comment]
> Hi all - please direct me to an issue link if there's a separate place for questions. We recently adopted Superset and I'm looking into setting up refresh schedules for our dashboards. This PR comes close to introducing the feature I think we need, but really something with the flexibility like cron schedule expressions would be ideal. I haven't been able to understand when the refresh interval triggers (i.e. how it's determined when it'll happen) or to specify a time it should kick off.
> 
> Effectively I just want to specify when my refresh interval should start and (based off the time window I choose) when it should stop. Can any help me understand this, please?

Hi there,
cron refresh would be an ideal solution to tell which dashboard should refresh all their pages at which time, based on the refresh schedule of underneath dataset.
Hi @rusackas I wonder is it ideal similar to the cache warm up feature, could you please take a look at this idea or fix the cache warm up feature, so that users will always be able to view the lastest update data without clicking into 'Force refresh' on charts or 'Refresh dashboard' on top.
Thank you

**Comment by rusackas** [comment]
After some quick sanity tests on the ephemeral, this is looking pretty good!

**Review by eschutho — APPROVED** [review]
LGTM! Thanks @Abhishek-kumar-samsung

**Comment by Abhishek-kumar-samsung** [comment]
> > Hi all - please direct me to an issue link if there's a separate place for questions. We recently adopted Superset and I'm looking into setting up refresh schedules for our dashboards. This PR comes close to introducing the feature I think we need, but really something with the flexibility like cron schedule expressions would be ideal. I haven't been able to understand when the refresh interval triggers (i.e. how it's determined when it'll happen) or to specify a time it should kick off.
> > Effectively I just want to specify when my refresh interval should start and (based off the time window I choose) when it should stop. Can any help me understand this, please?
> 
> Hi there, cron refresh would be an ideal solution to tell which dashboard should refresh all their pages at which time, based on the refresh schedule of underneath dataset. Hi @rusackas I wonder is it ideal similar to the cache warm up feature, could you please take a look at this idea or fix the cache warm up feature, so that users will always be able to view the lastest update data without clicking into 'Force refresh' on charts or 'Refresh dashboard' on top. Thank you

**Comment by Abhishek-kumar-samsung** [comment]
I accidently closed the PR and I reopened it again.

**Comment by rusackas** [comment]
@geido can you check your requested changes to (potentially) unblock the merge when you have a chance, please?

**Review by geido — APPROVED** [review]

**Comment by geido** [comment]
@Abhishek-kumar-samsung it seems that CI is stuck. Please pull latest master and push back to kick CI on a fresh master state. Thank you

**Comment by Abhishek-kumar-samsung** [comment]
> @Abhishek-kumar-samsung it seems that CI is stuck. Please pull latest master and push back to kick CI on a fresh master state. Thank you

Hi @geido i synced my fork to latest main branch, i don't know why the pull_request check is failing. 
From code side it did not gave any check failure.

Last time when it ran, the same code passed all checks.

**Comment by rusackas** [comment]
Thanks @Abhishek-kumar-samsung - I/we appreciate you sticking with this and getting it through!

**Comment by Abhishek-kumar-samsung** [comment]
> Thanks @Abhishek-kumar-samsung - I/we appreciate you sticking with this and getting it through!

Thankyou @rusackas

**Pull request merged**
